src/editable_tree/mod.rs

```diff
@@ -190,56 +322,113 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
         // The last value of nodes_to_clone is the node under the cursor, which we do not need to
         // clone, so we pop that reference.
         assert!(nodes_to_clone.pop().is_some());
-        self.finish_edit(&nodes_to_clone, new_node);
+
+        /* FINISH THE EDIT AND RETURN SUCCESS */
+
+        // Store the new_node's display name before it's consumed by `finish_edit`
+        let new_node_name = new_node.display_name();
+        self.finish_edit(&nodes_to_clone, 0, new_node);
+        EditResult::Ok(EditSuccess::Replace {
+            c,
+            name: new_node_name,
+        })
     }
 
-    /// Updates the internal state so that the tree now contains `new_node` inserted as the first
+    /// Updates the internal state so that the tree now contains `new_node` inserted as the last
     /// child of the selected node.  Also moves the cursor so that the new node is selected.
-    pub fn insert_child(&mut self, new_node: Node) -> Result<(), Node::InsertError> {
+    pub fn insert_child(&mut self, c: char) -> EditResult {
+        /* CHECK THE VALIDITY OF ARGUMENTS */
+
+        // Cache the node under the cursor, since finding the cursor involves non-trivial amounts
+        // of work
+        let cursor = self.cursor();
+        // Short circuit if `c` couldn't be a valid child of the cursor
+        if !cursor.is_insert_char(c) {
+            return EditResult::Err(EditErr::CannotInsertInto {
+                c,
+                parent_name: cursor.display_name(),
+            });
+        }
+
+        /* PERFORM THE ACTION */
+
+        // We know that `c` corresponds to a valid node (because `is_replace_char` returned
+        // `true`), so we can unwrap
+        let new_node = self.arena.alloc(cursor.from_char(c).unwrap());
         // Generate a vec of pointers to the nodes that we will have to clone.  We have to store
         // this as a vec because the iterator that produces them (cursor_path::NodeIter) can only
         // yield values from the root downwards, whereas we need the nodes in the opposite order.
         let mut nodes_to_clone: Vec<_> = self.current_cursor_path.node_iter(self.root()).collect();
-        let new_child_node = self.arena.alloc(new_node);
         // Clone the node that currently is the cursor, and add the new child to the end of its
         // children.  Unwrapping here is fine, because `cursor_path::NodeIter` will always return
         // one value.
         let mut cloned_cursor = nodes_to_clone.pop().unwrap().clone();
+        // Store the new_node's display name before it's consumed by `finish_edit`
+        let new_node_name = new_node.display_name();
         // Add the new child to the children of the cloned cursor
-        cloned_cursor.insert_child(new_child_node, cloned_cursor.children().len())?;
-        self.finish_edit(&nodes_to_clone, cloned_cursor);
-        Ok(())
+        cloned_cursor.insert_child(new_node, self.arena, cloned_cursor.children().len());
```

Comment:
I'm not sure what to do about these.  They return `Result<(), Node::InsertError>` which should be used, but as far as I remember, encapsulating errors in other errors is not recommended in Rust (I guess that it gets quite ugly...)

I see three options to fix this problem:
1. Require that `Node::InsertError` implements `Into<EditErr>` and provide some convenient things for them to convert into.
2. I can't see a way that insertion can fail other than the node having a fixed number of children (e.g. `true` has exactly 0 children, a `field` (`"key": "JSON"`) has exactly 2 children, etc...).  Likewise for deletion, except that deletion could also delete a child that doesn't exist (tbh, deleting a non-existent child is probably a cause for `panic!`).  So therefore, we could possibly just hardcode the error types of `Ast::insert_child` and `Ast::delete_child`, and then implement `From<InsertError>` and `From<DeleteError>` for `EditErr`. 
3. We just wrap the error anyway - have an enum value `Other(String)` that just logs the `String`, and convert `Ast::*Error` into that.

Reply:
> encapsulating errors in other errors is not recommended in Rust

Where did you hear such thing? It's actually quite opposite! Encapsulating errors is not just usual, the `std::error::Error` trait even has a method (`source()`) to deal with this. `thiserror` crate can smoothly help with that as well. I use this feature quite heavily.

There's just one thing to keep in mind: `Display` only displays outer error and the correct thing to do is to iterate over `source()` and display the inner errors formatted as you wish. The newest version of `slog` crate has a direct support for this (my contribution :smile:), you can see an [example of iterating](https://github.com/slog-rs/slog/blob/master/src/lib.rs#L2756) there. (The old `cause()` method is used because `slog` wants to support older versions of Rust.)

1. makes sense to implement conversion either way
2. fixed number of children is just a special case of range of valid numbers of children. Consider method in Rust trait declaration - it can have body, but doesn't have to and it can't have more than one body.
3. I'd recommend avoiding `String`, it'll make formatting hell.

Reply:
> Where did you hear such thing?
> ...
> 
> 3. I'd recommend avoiding `String`, it'll make formatting hell.

I can't remember - I think the Rust Book said something about it, but now I look for it I can't find anything...  I wonder if the Rust Book was actually warning me against encapsulating `String`s and I just didn't understand :laughing:.

> 2. fixed number of children is just a special case of range of valid numbers of children. Consider method in Rust trait declaration - it can have body, but doesn't have to and it can't have more than one body.

I think for now, I'm going to take this strategy - it'll probably have to be refactored again later (I also think it's generally good code hygiene to keep refactoring your code), but for now it'll get this PR ready so that other things like #32 can be done without merge conflicts.

As far as Rust trait definitions go, I would argue that perhaps Sapling should treat `trait T {}` and `trait T;` as both representing trait definitions with no `fn`s - i.e. a trait AST node would have >=0 `fn`s as its children rather than 0-1 bodies.  But arguing about this feels very hypothetical when Sapling can't even edit JSON yet :laughing:.

Reply:
(Finally) done!